go/vt/mysqlctl/azblobbackupstorage/azblob.go

```diff
@@ -0,0 +1,297 @@
+	return azblob.NewServiceURL(*u, pipeline)
+}
+
+// AZBlobBackupHandle implements BackupHandle for Azure Blob service.
+type AZBlobBackupHandle struct {
+	bs        *AZBlobBackupStorage
+	dir       string
+	name      string
+	readOnly  bool
+	waitGroup sync.WaitGroup
+	errors    concurrency.AllErrorRecorder
+}
+
+// Directory implements BackupHandle.
+func (bh *AZBlobBackupHandle) Directory() string {
+	return bh.dir
+}
+
+// Name implements BackupHandle.
+func (bh *AZBlobBackupHandle) Name() string {
+	return bh.name
+}
+
+// AddFile implements BackupHandle.
+func (bh *AZBlobBackupHandle) AddFile(ctx context.Context, filename string, filesize int64) (io.WriteCloser, error) {
+	if bh.readOnly {
+		return nil, fmt.Errorf("AddFile cannot be called on read-only backup")
+	}
+	// Error out if the file size it too large ( ~4.75 TB)
+	if filesize > azblob.BlockBlobMaxStageBlockBytes*azblob.BlockBlobMaxBlocks {
+		return nil, fmt.Errorf("filesize is too large to upload to az blob (max size %v)", azblob.BlockBlobMaxStageBlockBytes*azblob.BlockBlobMaxBlocks)
+	}
+
+	obj := objName(bh.dir, bh.name, filename)
+	containerURL, err := bh.bs.containerURL()
+	if err != nil {
+		return nil, err
+	}
+
+	blockBlobURL := containerURL.NewBlockBlobURL(obj)
+
+	reader, writer := io.Pipe()
+	bh.waitGroup.Add(1)
+
+	go func() {
+		defer bh.waitGroup.Done()
+		_, err := azblob.UploadStreamToBlockBlob(ctx, reader, blockBlobURL, azblob.UploadStreamToBlockBlobOptions{
+			BufferSize: azblob.BlockBlobMaxStageBlockBytes,
+			MaxBuffers: *azBlobParallelism,
+		})
+		if err != nil {
+			reader.CloseWithError(err)
+			bh.errors.RecordError(err)
+		}
+	}()
+
+	return writer, nil
+}
+
+// EndBackup implements BackupHandle.
+func (bh *AZBlobBackupHandle) EndBackup(ctx context.Context) error {
+	if bh.readOnly {
+		return fmt.Errorf("EndBackup cannot be called on read-only backup")
+	}
+	bh.waitGroup.Wait()
+	return bh.errors.Error()
+}
+
+// AbortBackup implements BackupHandle.
+func (bh *AZBlobBackupHandle) AbortBackup(ctx context.Context) error {
+	if bh.readOnly {
+		return fmt.Errorf("AbortBackup cannot be called on read-only backup")
+	}
+	return bh.bs.RemoveBackup(ctx, bh.dir, bh.name)
+}
+
+// ReadFile implements BackupHandle.
+func (bh *AZBlobBackupHandle) ReadFile(ctx context.Context, filename string) (io.ReadCloser, error) {
+	if !bh.readOnly {
+		return nil, fmt.Errorf("ReadFile cannot be called on read-write backup")
+	}
+
+	obj := objName(bh.dir, bh.name, filename)
+	containerURL, err := bh.bs.containerURL()
+	if err != nil {
+		return nil, err
+	}
+	blobURL := containerURL.NewBlobURL(obj)
+
+	resp, err := blobURL.Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body(azblob.RetryReaderOptions{
+		MaxRetryRequests: defaultRetryCount,
+		NotifyFailedRead: func(failureCount int, lastError error, offset int64, count int64, willRetry bool) {
+			bh.errors.RecordError(lastError)
```

Comment:
Recording an error here means the backup will be considered as failed. I don't think we want to do that since the docs for `NotifyFailedRead` say it gets called "after any failure to read". That seems to imply it will be called even on failures that get retried and later succeed.

Reply:
I moved this to also use willRetry to only flag the error when we won't retry it. I am going to log at info when it fails but we are trying